lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs

```diff
@@ -286,6 +251,59 @@ public void OnNext(ICompletedTask completedTask)
             }
         }
 
+        public void OnNext(IFailedEvaluator value)
+        {
+            Logger.Log(Level.Info, "An evaluator failed, checking if it failed before context and service was submitted");
+            int currFailedEvaluators = Interlocked.Increment(ref _currentFailedEvaluators);
+
+            if (value.FailedTask != null && value.FailedTask.IsPresent())
+            {
+                Exceptions.Throw(new Exception("An evaluator failed during task, cannot continue IMRU task"), Logger);
+                Logger.Log(Level.Error, "An evaluator failed during task, cannot continue IMRU task");     
+            }
+
+            if (value.FailedContexts != null && value.FailedContexts.Count != 0)
+            {
+                Exceptions.Throw(new Exception("Some active context failed, cannot continue IMRU task"), Logger);
+                Logger.Log(Level.Info, "Some active context failed, cannot continue IMRU task");        
+            }
+
+            if (currFailedEvaluators > _allowedFailedEvaluators)
+            {
+                Exceptions.Throw(new Exception("Cannot continue IMRU job, Failed evaluators reach maximum limit"),
+                    Logger);
+            }
+
+            Logger.Log(Level.Info, "Requesting for the failed evaluator again");
+
+            IEvaluatorRequest request = null;
+            _serviceAndContextConfigurationProvider.EvaluatorFailed(value.Id);
+
+            //If active context stage is reached for Update Task then assume that failed
+            //evaluator belongs to mapper
+            if (_reachedUpdateTaskActiveContext)
+            {
+                request =
```

Comment:
Maybe make a helper method that only accepts the number of Evaluators to request? We have this same code in several places now that fault tolerance is added. E.g. `private void RequestMapEvaluators(int number){}`